uefi/src/table/system.rs

```diff
@@ -146,73 +146,84 @@ impl SystemTable<Boot> {
-    /// for the memory map right before exiting boot services, and to allocate a
-    /// bit more storage than requested by memory_map_size.
-    ///
-    /// If `exit_boot_services` succeeds, it will return a runtime view of the
-    /// system table which more accurately reflects the state of the UEFI
-    /// firmware following exit from boot services, along with a high-level
-    /// iterator to the UEFI memory map.
+    /// This function panics if it is unable to allocate memory to hold the
+    /// memory map. It will also panic if, after a few retries, it is unable to
+    /// retrieve the current memory map needed to exit boot services.
     ///
     /// [`global_allocator::exit_boot_services`]: crate::global_allocator::exit_boot_services
     /// [`Logger::disable`]: crate::logger::Logger::disable
     /// [`uefi_services::init`]: https://docs.rs/uefi-services/latest/uefi_services/fn.init.html
-    pub fn exit_boot_services(
-        self,
-        image: Handle,
-        mmap_buf: &mut [u8],
-    ) -> Result<(SystemTable<Runtime>, MemoryMapIter<'_>)> {
-        unsafe {
-            let boot_services = self.boot_services();
-
-            loop {
-                // Fetch a memory map, propagate errors and split the completion
-                // FIXME: This sad pointer hack works around a current
-                //        limitation of the NLL analysis (see Rust bug 51526).
-                let mmap_buf = &mut *(mmap_buf as *mut [u8]);
-                let mmap_comp = boot_services.memory_map(mmap_buf)?;
-                let (mmap_key, mmap_iter) = mmap_comp;
-
-                // Try to exit boot services using this memory map key
-                let result = boot_services.exit_boot_services(image, mmap_key);
-
-                // Did we fail because the memory map was updated concurrently?
-                if result.status() == Status::INVALID_PARAMETER {
-                    // If so, fetch another memory map and try again
-                    continue;
-                } else {
-                    // If not, report the outcome of the operation
-                    return result.map(|_| {
-                        let st = SystemTable {
-                            table: self.table,
-                            _marker: PhantomData,
-                        };
-                        (st, mmap_iter)
-                    });
+    #[must_use]
+    pub fn exit_boot_services(self) -> (SystemTable<Runtime>, MemoryMapIter<'static>) {
+        // Arbitrarily limit the number of attempts to a reasonable number. If
+        // it doesn't succeed in this number of attempts, more attempts are
+        // unlikely to help.
+        let max_attempts = 10;
+
+        let boot_services = self.boot_services();
+        let mut extra_entries = 2;
+
+        for _ in 0..max_attempts {
+            let memory_map_size = self.boot_services().memory_map_size();
+
+            // Get a buffer size that should be able to hold the memory
```

Comment:
Which kind of extra allocations do we try to work-around here?
I think we should instead try to have a fine-grained control over which allocations may happen in this specific code path

Reply:
I don't think it's possible precisely control which allocations occur here, because a higher-priority task could be executed via timer. This [table](https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html#tpl-restrictions) shows restrictions on what priority level various functions can be called at. `ExitBootServices` must be called at the application priority level, which is the lowest level. So we can't change what TPL this code is running at.

We also don't control the implementation of `allocate_pool`. Intuitively I wouldn't expect that allocation to increase the number of memory map entries by more than one, but I don't think that's is guaranteed by the spec.

In general I would expect this loop to succeed on the very first try, but there's a small chance it won't.

Reply:
I don't know if it's valid, but we can disable interrupts for a short period of time.. For the scope of this function. What do you think?

Reply:
I'm not sure if disabling interrupts is valid -- it's not explicitly mentioned anywhere in the spec that I have seen.

I spent some time thinking more about the problem, and also looking into how it's handled in the Linux kernel since that provides some better proof of how stuff works on real firmware out in the wild, not just OVMF in QEMU.

One interesting thing I learned is that the spec for exit boot services changed slightly starting in UEFI 2.9. In UEFI 2.8 and earlier, only `get_memory_map` and `exit_boot_services` may be called after the first call to `exit_boot_services`. Starting in UEFI 2.9, other memory allocation functions may also be called.

The way the Linux kernel handles the memory allocation is to add space for 8 extra entries: https://github.com/torvalds/linux/blob/e544a07438/drivers/firmware/efi/libstub/efistub.h#L173. If exit boot services fails the first time, they try it only one more time: https://github.com/torvalds/linux/blob/e544a0743/drivers/firmware/efi/libstub/efi-stub-helper.c#L375. I've updated this PR to also use that constant and retry-once-behavior.

In the Linux kernel, if exit boot services fails after the second time, they end up calling `boot_services.exit` to return control to the firmware (or to another boot services program that launched the current one). I think technically that's against the spec since `exit` isn't one of the allowed boot services functions to call after exit_boot_services fails. So in this PR I've changed it to just `reset()` the system. I think that's probably the right behavior to have since the system is pretty much in an unknown and unrecoverable situation otherwise. And exiting boot services is considered a security boundary as well, so I think any other action would run the risk of introducing a secure boot bypass.

I think the most important thing is to make the implementation reasonably robust so that it doesn't fail in the first place. Exit boot services really isn't supposed to fail as long as you've retrieved the memory map properly, so all other failure handling is really an exceptional panic situation.

Reply:
Thanks for the investigation! 🤘